[Feat] Button 컴포넌트 생성

aken-you opened the pull request:
# 관련 이슈 번호

close #17 

# 설명

- colorType: 주요 색상을 결정
- variant: 스타일
- size: 사이즈
- disabled: 비활성화 여부

<img width="607" alt="스크린샷 2024-08-20 오후 3 46 12" src="https://github.com/user-attachments/assets/037c6e44-b17d-485f-aa33-a9b43569723d">


# 첨부 파일 (Optional)

# 레퍼런스 (Optional)

aken-you committed:
feat[#17] : Button 컴포넌트 공통 스타일 적용

aken-you committed:
feat[#17] : Button 컴포넌트 색상 스타일 적용

aken-you committed:
feat[#17] : Button 컴포넌트 사이즈 스타일 설정

- size에 따라 padding, text size, height가 달라짐
- icon, text의 유무에 따라 padding이 달라짐

aken-you committed:
refactor[#17] : Button 컴포넌트 색상 스타일 수정

- type이 text이고 state가 enabled이면, background-color는 없음

aken-you committed:
refactor[#17] : Button 컴포넌트의 스타일을 정의하는 파일 추가

aken-you committed:
refactor[#17] : Button 컴포넌트 outline: none으로 설정

aken-you committed:
style[#17] : children의 첫 번째 자식 요소를 저장하는 변수와 마지막 자식 요소를 저장하는 변수 네이밍 변경

aken-you committed:
refactor[#17] : 첫 번째 자식 요소의 타입이 string인지 저장하는 변수 생성

aken-you committed:
refactor[#17] : 마지막 자식 요소의 타입이 string인지 저장하는 변수 생성

aken-you committed:
refactor[#17] : Button 컴포넌트가 disabled 상태일 때, 포인터 이벤트(클릭, 호버, 포커스 등)를 비활성화

aken-you committed:
refactor[#17] : :focus 가상 클래스를 :focus-visible로 변경

aken-you committed:
feat[#17] : Button 컴포넌트 스토리북 작성

aken-you committed:
chore[#17] : tailwind.config.js에서 오타 수정

aken-you committed:
feat[#17] : Button 컴포넌트 export

yonghyeun commented:
잘 읽었습니다 ~~ 끝내주게 깔끔하고 멋진 코드네용 

저도 코드에 주석처리를 좀 달아서 더 이해하기 쉬운 코드를 쓰도록 노력해봐야겠습니다

yonghyeun commented on src/shared/ui/button/Button.tsx:
```diff
@@ -0,0 +1,67 @@
+  const baseStyles =
+    "inline-flex flex-shrink-0 items-center justify-center gap-[.625rem] rounded-[1rem] outline-none";
+  const colorStyles = colors[colorType][variant];
+
```
휴먼에러를 방지하기 위해 chidlren.length가 디자인 시스템 조건을 만족하지 않을 경우 분기처리를 해보면 어떨까요 ? 🔥

yonghyeun reviewed (commented):
친절한 주석처리와 깔끔한 코드까쥐 .. 잘 보고 갑니다

aken-you committed:
feat[#17] : Button 컴포넌트의 children 타입 설정

- ButtonHTMLAttributes<HTMLButtonElement>에서 children이 필수가 아니다.
- children을 필수값으로 입력하기 위해, children 타입 설정

aken-you committed:
refactor[#17] : children에 여러 요소가 있을 경우를 대비해 조건 수정

- 이전에는 children에 올 수 있는 하위 요소를 2개만 허락했었음
- children에 여러 요소가 있더라도 스타일이 적용될 수 있도록 조건 수정

aken-you commented on src/shared/ui/button/Button.tsx:
```diff
@@ -0,0 +1,71 @@
+  colorType: "primary" | "secondary" | "tertiary";
+  variant: "filled" | "outlined" | "text";
+  size: "xSmall" | "small" | "medium" | "large";
+  children: ReactNode;
```
기존에 children 입력이 필수가 아니었어요.
그래서 타입을 수정하여, children을 필수값으로 설정했습니다 :)

yonghyeun replied:
커어어억 감사합니다 확인했습니다

aken-you commented on src/shared/ui/button/Button.tsx:
```diff
@@ -0,0 +1,71 @@
+  const hasOnlyText = hasOneChild && isFirstChildText;
+
+  // children에 있는 요소가 여러개인지 확인
+  const hasMultipleChildren = childrenArray.length >= 2;
```
children에 2개 이상의 요소가 있으면 스타일이 적용되도록, 조건을 수정했습니다!

yonghyeun replied:
children 에 2가지 초과의 요소가 있어도 에러 처리를 하지 않는 이유는 무엇인지 여쭤봐도 될까요 ?

aken-you replied:
`childrenArray.length === 2`
이전에는 두개만 허용하여, 2개를 초과할 경우 스타일이 적용되지 않았어요.

그래서 2개를 초과해도 스타일이 깨지지 않도록 하여, 변경에 유연한 컴포넌트를 만들고 싶었어요.

yonghyeun replied:
확인했습니다 :)

aken-you committed:
style[#17] : Button 컴포넌트 props 네이밍 변경

yonghyeun commented on src/shared/ui/button/Button.tsx:
```diff
@@ -0,0 +1,77 @@
+    (isValidElement(lastChild) && lastChild.type === "span") ||
+    (isValidElement(lastChild) && lastChild.type === "p");
+
+  const hasIconFirst = hasMultipleChildren && isLastChildText;
```
코드만 보았을 때 children 요소가 2개 이상이면서 [text  , text] 일 경우에도 `hasIconFirst` 가 true 가 될 것으로 예상되는데 확인 부탁드립니다.

yonghyeun reviewed (commented):
제가 워낙 휴먼에러를 자주 내는 얼렁뚱땅이라 해당 부분에 중점을 두고 리뷰해보았습니다

aken-you committed:
feat[#17] : Button 컴포넌트의 children에 여러 개의 텍스트 요소를 포함할 경우 error throw

yonghyeun commented on src/shared/ui/button/Button.tsx:
```diff
@@ -0,0 +1,82 @@
+  const lastChild = childrenArray[childrenArray.length - 1];
+  const isLastChildText = isText(lastChild);
+
+  const areAllChildrenText = childrenArray.every(isText);
```
isText 메소드와 배열을 이용한 방법이 세련되고 상당히 멋진 거 같습니다 🔥

yonghyeun commented on src/shared/ui/button/Button.tsx:
```diff
@@ -0,0 +1,82 @@
+    );
+  }
+
+  const hasIconFirst = hasMultipleChildren && isLastChildText;
```
`hasIconFirst` 에서 조건문에서 첫 번째 요소가 아이콘임을 판단하는 기준이 마지막 요소의 타입이 text 요소인지 확인하는 것 보다 첫 번째 요소가 text 요소가 아님으로 하는 거는 어떨까요 ? 

children 요소의 최대 개수를 2개로 엄격히 지정하는 경우엔 해당 조건문이 깨지지 않지만 유연성을 위해 children의 최대 개수를 열어 놓는 경우 [icon, text,icon] 일 경우에는 해당 조건문이 깨질 거 같아 코멘트 남겨봅니다.

aken-you replied:
[icon, text,icon]일 경우 padding 값이 좌측은 16px, 우측은 24px이라서 어색할 거 같네요 👀 
차라리 child를 2개로 제한해서, 2개 초과인 경우 오류를 내도록 구현하는게 더 좋아보이네요

yonghyeun reviewed (commented):
빠른 피드백 감사합니다 !

aken-you committed:
refactor[#17] : Button 컴포넌트의 여러 개의 자식이 있는 경우에 관한 조건 수정

- 2개를 초과할 경우 에러 처리
- hasIconFirst와 hasTextFirst 조건 수정

yonghyeun commented on src/shared/ui/button/Button.tsx:
```diff
@@ -0,0 +1,80 @@
+  // children에 있는 요소가 여러개인지 확인
+  const haveTwoChildren = childrenArray.length === 2;
+
+  if (childrenArray.length > 2) {
```
저희 디자인 시스템에선 children 요소들은 필수적으로 서로 다른 타입을 가져야 하는 것으로 보입니다. 

- icon
- text
- icon , text
- text , icon

해당 분기처리가 들어가면 어떨까요 ? 

현재의 분기문에선 throw 하는 에러 메시지를 전부 캐치하지 못하지 않을까 생각됩니다

aken-you replied:
>현재의 분기문에선 throw 하는 에러 메시지를 전부 캐치하지 못하지 않을까

여기서 에러가 일어날 수 있는 경우는  [icon , icon] , [text ,text] 밖에 없다고 생각하는데 맞을까요?!
지금 처리된 분기처리가 말씀하신 4가지로 되어있어서요

yonghyeun replied:
네 맞습니다 지금 코드로 제가 발생 시킬 수 있는 휴먼에러가 모두 처리 된 거 같습니다 호호

yonghyeun commented on src/shared/ui/button/Button.tsx:
```diff
@@ -0,0 +1,80 @@
+  const lastChild = childrenArray[childrenArray.length - 1];
+  const isLastChildText = isText(lastChild);
+
+  const hasIconFirst = haveTwoChildren && !isFirstChildText && isLastChildText;
```
상단 분기문 이후 firstChild , lastChild 의 타입이 서로 다름이 보장된다면 `isFirstChild , isLastChild`  두 개의 조건문을 사용하지 않아도 될 것이라고 생각되는데 어떻게 생각하시나요 ?

현재의 분기문 이후 코드에선 여전히 [icon , icon] , [text ,text]의 경우를 걸러내지 못하지 않나 싶습니다.   👁️👁️

aken-you replied:
firstChild와 lastChild의 타입이 다른 경우를 생각해서 넣은 거였는데, [icon , icon] , [text ,text]의 경우를 처리하지 않았네요!

yonghyeun reviewed (commented):
빠른 피드백 감사합니다. 에러 분기문 처리 관련하여 리뷰 남깁니다

aken-you committed:
refactor[#17] : Button 컴포넌트에서 children 타입이 동일할 경우 에러 처리

yonghyeun commented on src/shared/ui/button/Button.tsx:
```diff
@@ -0,0 +1,90 @@
+  const hasIconFirst = haveTwoChildren && !isFirstChildText && isLastChildText;
+  const hasTextFirst = haveTwoChildren && isFirstChildText && !isLastChildText;
+
+  const isSameType =
```
해당 조건문은 `isFirstChildText && isLastChildText` 만 있어도 되지 않을까요 ? `||` 후에 존재하는 조건문은 이전과 동일한 결과를 뱉는 거 같습니다 !!

aken-you replied:
1. `isFirstChildText && isLastChildText` => [text, text]
2. `!isFirstChildText && !isLastChildText` => [icon, icon]

이 두가지를 고려해서 짰어요
`||` 후에 존재하는 2번 코드는 [icon, icon]을 고려한 부분이었어요!

yonghyeun replied:
확인했습니다 :)

yonghyeun reviewed (commented):
마지막 리뷰일 거 같습니다 완벽하고 좋네요 멋집니다 !!

aken-you committed:
refactor[#17] : isSameType 위치 변경

aken-you committed:
feat[#17] : Button 컴포넌트에 forwardRef 적용

aken-you committed:
refactor[#17] : index 확장자 tsx -> ts로 변경
